docs/design.md

```diff
@@ -0,0 +1,118 @@
+
+The `setReceiver` method is called by the bftengine code to let implementers of
+the `ICommunication` interface know where received messages should be handled.
+Handlers for received messages implement the `IReceiver` interface. The two sole
+implementers of `IReceiver` are the [ReplicaImp](../bftengine/src/bftengine/ReplicaImp.hpp)
+and the [SimpleClient](../bftengine/src/bftengine/SimpleClient.cpp)
+
+## State transfer
+
+State transfer refers to the mechanism via which out of date replicas can
+retrieve already committed data efficiently and bring themselves in sync with
+the rest of the cluster. State transfer is done outside of the ordering and
+replication protocol performed by SBFT, and can often be achieved by data transfer
+between only two replicas.
+
+As state transfer is outside of the byzantine agreement protocol, it must be
+built as a function of the instantiation of the specific state machine
+implementation. Concord-bft treats state machine operations and data to be
+replicated/executed upon commit as opaque. Since the data being stored by the
+application is opaque, and can be arbitrarily large, concord-bft cannot provide
+a universally efficient mechanism for transferring such state, and relies on
+state machine specific implementations for optimal distribution.
+
+For applications to implement state transfer, and have concord-bft properly be
+able to initiate state transfer when needed, and also know when there is more
+state that needs to be transferred, there must be hooks in both the application
+code and concord-bft code that allow interaction between the application and
+bftengine in a generic manner. This is the purpose of the state transfer
+interfaces in
+[IStateTransfer.hpp](../bftengine/include/bftengine/IStateTransfer.hpp).
+
+The `IStateTransfer` interface is implemented by the application developer for a given
+state transfer/checkpoint mechanism. It is called by the main replica thread of
+the core bftengine implementation. The concord-bft library contains two
+implementations of `IStateTransfer`:
+[SimpleStateTransfer](../bftengine/src/simplestatetransfer) and
+[BcStateTransfer](../bftengine/src/bcstatetransfer). These will be described
+further in the detailed design section for state transfer.
+
+The `IReplicaForStateTransfer` interface allows the state transfer
+implementations to call functions in the bftengine replica.
+
+## Metadata Storage
+
+The bftengine has internal metadata that must be persisted to disk. The
+[MetadataStorage](../bftengine/include/bftengine/MetadataStorage.hpp) interface
+should be implemented by the application developer for a given storage system
+such as RocksDB.
+
+## Replica
+
+The [Replica](../bftengine/include/bftengine/Replica.hpp) interface is the API
+used to instantiate a new replica given specific implementations of the other
+public interfaces.
+
+## RequestsHandler
+
+The [RequestsHandler](../bftengine/include/bftengine/Replica.hpp) interface
```

Comment:
The RequestsHandler interface contains a single callback function to be implemented by the application. The bftengine uses the RequestsHandler interface when a given sequence number is committed.